### ISSUE TITLE
instance_seg dataset

### ISSUE THREAD

**Issue by xpwindells** [QuentinFitteRey/VLMSAM#2]
python file:cot_dataset.py
*****************************************************************************************************
        if "instance_seg" in low:
            conversations, masks = [], []
            for inst_id, class_name in instances:
                question = f"What is the {class_name} in this image? Please output segmentation mask."
                answer = "[SEG]"
                conv = get_default_conv_template(self.model_name).copy()
                user_msg = f"<|vision_start|><|image_pad|><|vision_end|>\n{question}"
                conv.append_message(conv.roles[0], user_msg)
                conv.append_message(conv.roles[1], answer)
                conversations.append(conv.get_prompt())

                rle_info = item.get("info", {}).get(inst_id)
                if item.get("is_rle") and isinstance(rle_info, dict) and "counts" in rle_info:
                    if not isinstance(rle_info["counts"], bytes):
                        rle_info["counts"] = rle_info["counts"].encode()
                    mask = mask_utils.decode(rle_info)
                    if mask.shape != (h, w):
                        mask = cv2.resize(mask, (w, h), interpolation=cv2.INTER_NEAREST)
                    masks.append(mask.astype(np.uint8))
                else:
                    polys = item.get("info", {}).get(inst_id)
                    if polys is not None:
                        masks.append(_fill_from_polys(polys))
                # If no valid mask, skip adding to masks list

            return conversations, masks
**********************************************************************************************************************
Hey，i found the mask was not correctly loaded, and the question did not use the dataset question published by Lisa++.Could you please explain the reason for doing this?

**Comment by QuentinFitteRey** [comment]
Hey I've been busy those past weeks, i will try to take a look on that but i'm currently working on a possibly better version and simpler. I will try to investigate but could you detail the mask not correctly loading?

For the fact of not using the question, i dont find it necessary to make the model better, a core limitation of those models is that the dataset part that allows to make it more conversational is really limited in application when you derived from easy examples.
Therefore, i just use a default prompt and get the right class name

**Comment by xpwindells** [comment]
Thank you for your reply. The JSON format of the instance_seg dataset in the official Lisa++ repository is inconsistent with the formats of cot, conversation, and caption. This has caused errors during the loading process. By the way, after training, the metrics on the validation set are normal, but the inference results obtained from the inference.py script are very poor. Is there any parameter that needs to be modified?

**Comment by QuentinFitteRey** [comment]
Okay that's where we aligned, I think that the way Lisa used to validate is wrong. You can find in the folder model, a version where validation is made with generation because this is important to test your model generating and not doing as they do (giving the text only computing loss and then giving the right text to SAM). Therefore, you should first try to validate with generation to veirfy you training is correct.

Secondly, for inference, you should respect prompt defined in the training. Look in utils to find prompt that will make the model understand its tasks. Also, this is a limitation and as I said I'm trying to work on a new method that will improve those results but its WIP

**Comment by xpwindells** [comment]
The multi-instance data requires a changelist to assist in matching the mask and the gt_mask. Why is the method Lisa used for validation incorrect?

**Comment by QuentinFitteRey** [comment]
Basically, they don’t generate the sentence; they directly use the ground truth sentence, apply a forward pass on it to get the hidden layer, and send it to SAM. I personally think that this isn’t the right way to validate, because when you generate your own sentence with the model, it might not produce the same one or could carry less information. That’s why I’ve been testing different evaluation methods, as in the other modeling file. However, with my personal weights, I get better evaluation metrics using their validation method, but it just feels wrong to me.

**Comment by xpwindells** [comment]
> Basically, they don’t generate the sentence; they directly use the ground truth sentence, apply a forward pass on it to get the hidden layer, and send it to SAM. I personally think that this isn’t the right way to validate, because when you generate your own sentence with the model, it might not produce the same one or could carry less information. That’s why I’ve been testing different evaluation methods, as in the other modeling file. However, with my personal weights, I get better evaluation metrics using their validation method, but it just feels wrong to me.

I understand your point. You want to use "generate" for verification as well. I also think it's inappropriate to use "forward" for val. However, my problem is that the performance of the qwenvl2.5 model is extremely poor during multi-instance segmentation, and this issue has been bothering me for a long time.
